adapters/repos/db/vector/multivector/muvera.go

```diff
@@ -0,0 +1,276 @@
+					for i := 0; i < e.config.Dimensions; i++ {
+						tmpVec[startIdx+uint64(i)] = fullVec[nearestPoint][i]
+					}
+				}
+			}
+		}
+		// doc ONLY operations ended
+
+		scale := 1.0 / float32(math.Sqrt(float64(e.config.DProjections)))
+		projOffset := rep * e.config.NumClusters * e.config.DProjections
+		matrix := e.S[rep]
+		// Process each cluster
+		for j := 0; j < e.config.NumClusters; j++ {
+			// Calculate source and destination offsets
+			srcStart := j * e.config.Dimensions
+			dstStart := projOffset + (j * e.config.DProjections)
+
+			// Process in chunks of 4 for better cache utilization
+			for k := 0; k < e.config.DProjections; k++ {
+				var sum float32
+				// Process 4 elements at a time
+				for l := 0; l < e.config.Dimensions; l += 4 {
+					end := l + 4
+					if end > e.config.Dimensions {
+						end = e.config.Dimensions
+					}
+					// Unroll the inner loop
+					for m := l; m < end; m++ {
+						sum += matrix[k][m] * tmpVec[srcStart+m]
+					}
+				}
+				encodedVec[dstStart+k] = sum * scale
+			}
+		}
+
+		// Reset tmpVec, this is needed only for query encoding
+		for i := range tmpVec {
+			tmpVec[i] = 0
+		}
+	}
+
+	return encodedVec
+}
+
+// EncodeQuery encodes a query vector using Muvera
+func (e *MuveraEncoder) EncodeQuery(query [][]float32) []float32 {
+	projectedQuery := e.encode(query, false)
+
+	return projectedQuery
+}
+
+// EncodeDoc encodes a document vector using Muvera
+func (e *MuveraEncoder) EncodeDoc(fullDoc [][]float32) []float32 {
+	projectedDoc := e.encode(fullDoc, true)
+
+	return projectedDoc
+}
+
+func MuveraBytesFromFloat32(vec []float32) []byte {
+	slice := make([]byte, len(vec)*4)
+	for i := range vec {
+		binary.LittleEndian.PutUint32(slice[i*4:], math.Float32bits(vec[i]))
+	}
+	return slice
+}
+
+func MuveraFromBytes(bytes []byte) []float32 {
+	vec := make([]float32, len(bytes)/4)
+	for i := range vec {
+		vec[i] = math.Float32frombits(binary.LittleEndian.Uint32(bytes[i*4 : (i+1)*4]))
+	}
+	return vec
+}
+
+func (e *MuveraEncoder) GetMuveraVectorForID(id uint64, bucket string) ([]float32, error) {
+	idBytes := make([]byte, 8)
+	e.storeId(idBytes, id)
+	muveraBytes, err := e.muveraStore.Bucket(bucket).Get(idBytes)
+	if err != nil {
+		return nil, fmt.Errorf("getting vector for id: %w", err)
+	}
+	if len(muveraBytes) == 0 {
+		return nil, fmt.Errorf("vector not found for id %d", id)
+	}
+
+	return MuveraFromBytes(muveraBytes), nil
+}
+
+type MuveraData struct {
+	KSim         uint32        // 4 bytes
+	NumClusters  uint32        // 4 bytes
+	Dimensions   uint32        // 4 bytes
+	DProjections uint32        // 4 bytes
+	Repetitions  uint32        // 4 bytes
+	Gaussians    [][][]float32 // 4 bytes -> (repetitions, numClusters, dimensions)
+	S            [][][]float32 // 4 bytes -> (repetitions, dProjections, dimensions)
+}
+
+type CommitLogger interface {
+	AddMuvera(MuveraData) error
+}
+
+func (e *MuveraEncoder) PersistMuvera(logger CommitLogger) {
+	logger.AddMuvera(MuveraData{
```

Comment:
We should handle the error here or bubble it up

Reply:
It's better to abort immediately in my opinion, so that the error displayed in the logs has all of context.